### ISSUE TITLE
lfe-compile needs updating for newest rebar3

### ISSUE THREAD

**Issue by timclassic** [lfe-rebar3/compile#1]
I get stack traces when trying to use lfe-compile with a recent rebar3.  I did not have time to port it myself, but I got far enough to notice the following API changes:
- `rebar_app_info:state_or_new/2` no longer exists.
- `rebar_dir:all_src_dirs/3` now takes a `rebar_dict()` instead of a `rebar_app_info:t()`.  @tsloughter says that some info was moved from rebar_state to app_info.  `rebar_app_info:opts/1` can be used to pull the field out of the app_info.

There may be more.  I hope this info helps to bootstrap the effort!

**Comment by oubiwann** [comment]
I'll give it a look -- thanks!

**Comment by oubiwann** [comment]
Hrm. I may need an example project that you're having a problem with to run against -- I wasn't able to find a problem.

Here's what I did:
- `git clone git@github.com:rebar/rebar3.git`
- `cd rebar3 && ./bootstrap && sudo mv rebar3 /usr/local/bin`
- `lfetool new library lfesample`
- `cd lfesample`
- I then updated the project's `rebar.config` file to include the plugin, per the instructions in the rebar3-lfe-compile README
- `rebar3 compile`

There were lots of warnings from LFE itself (e.g., the standard shadowing warnings), but the final lines were this:

```
===> Compiling ltest
===> Compiling lutil
===> Compiling lfesample
```

Just to make sure, I did the following:

```
$ rebar3 shell
1> lfe_shell:server().
> (+ 6 6 7 7 8 8)
42
```

Let me know if there are some steps I've overlooked ...

**Comment by oubiwann** [comment]
Man, it has been way too long since I've looked at this. I didn't even run `rebar3 lfe compile`. I see the problem now :-/

Looking into it ... again ...

**Comment by oubiwann** [comment]
Hacked on this some tonight (in its own branch, https://github.com/lfex/rebar3-lfe-compile/tree/use-new-rebar3-compile). Fixed a bunch of things and updated the plugin code to use the new rebar3 code. However, after fixing all the errors, I'm now getting a rather puzzling `undef` error. I'll take another look at it when I'm not falling asleep at the keyboard, but here's the latest:
- https://gist.githubusercontent.com/oubiwann/0a624d06d736ea8914c6/raw/409866a8c295ce3ff3a740e5cc827a6ab822de01/rebar3_lfe_compiler_error.erl

**Comment by oubiwann** [comment]
Ah, I see why it's giving the `undef`, now that I've had some sleep: the LFE compiler is not compiling the deps in `_build/default/deps` before trying to compile the code in the lfesample project. The rebar3 LFE compiler does actually compile `src/lfesample.lfe`, but it barfs on `src/lfesample-util.lfe` because that one depends upon `lutil` which is not compiled in in the deps.

The code that _should_ compile the deps does seem to be in place, so there is something not-quite-so-obvious at work here. Will have to dig tonight after work.

**Comment by oubiwann** [comment]
Oh, holy crap -- plugin development for custom compilers has _completely_ changed:
- https://www.rebar3.org/docs/custom-compiler-plugins

This is actually the sort of thing I was asking the rebar3 devs for -- so this is great to see. Took me too long to find it, though :-/

**Comment by timclassic** [comment]
@oubiwann I didn't expect to trigger such activity :)  I am following your progress, and please let me know if you'd like me to test anything.

**Comment by oubiwann** [comment]
@timclassic Will do! And no worries about the activity :-) There've actually been some subtle issues with the rebar3 LFE compiler (e.g., dependencies of plugins not getting compiled) so it has been on my list to address this. One of the big blockers for continued development was the very painful lack of custom-compiler support in rebar3. I basically had to copy a bunch of the internal functions from rebar3's Erlang compiler, while making bug-generating changes to remove uncessary bits and support LFE's differences. It was quite a mess.

Since then, the rebar3 team has been quite busy and have not only made code changes to better support these scenarios, but they've provided docs! I was quite pleased last night when I discovered this.

All in all, the time is ripe to finally get LFE support to 100% with rebar3. As long as I can still hack on it before bedtime :-)

I'm hitting other snags as I hack away on it, but they're slowly getting worked out. I'm hopeful there will be something to test soon ...

**Comment by timclassic** [comment]
@oubiwann Thanks for the details.

FYI, my desired use case is to be able to build a project with rebar3 that has both Erlang code and LFE code, and that code may be in `src/`, `apps/*/src`, or a dependency.  Ideally it should all Just Work as long as I have the rebar3/LFE integration set up as a plugin.

How do you see such a configuration playing out?  Will I need to run `rebar3 compile` and then `rebar3 lfe compile` to build everything?

**Comment by oubiwann** [comment]
Yeah, that should work when this is all done. The changes that the rebar team made means that you will only have to run `rebar3 compile` -- as long as you have your `rebar.config` file set up properly (I'll need to make changes to lfetool so that new projects have the proper `rebar.config` file setup). 

The rebar3-lfe-compile README will have instructions for folks who want to use it and who didn't generate their projects with lfetool (iow, it will show they what changes they need to make to most effectively take advantage of the plugin).

**Comment by timclassic** [comment]
@oubiwann Sounds great!

**Comment by oubiwann** [comment]
Success! Turned out the signatures for `rebar_base_compiler:ok_tuple` and `rebar_base_compiler:error_tuple``, and that was the last thing I needed for the compiles to finally work after switching over to the new methodology.

There's still some minor stuff I need to fix up (lfe itself is getting compiled twice), but it should be ready to test now. I'll update the docs, but in order to work, the following must be done:
1. Your project's `rebar.config` file needs to declare the rebar3-lfe-compile plugin dep
2. Your projects's `rebar.config` file needs to declare a provider hook
3. All of your project's LFE dependencies need to do the same

I haven't updated the core LFE libraries to use rebar3 yet ... that's going to change things in a big way, since rebar2 doesn't seem to work with the updated `rebar.config` files when they've been updated to use the new LFE compiler plugin. I'll probably send out an email to the mail list tonight, giving a heads up, etc.

So, here's an example of what your LFE project `rebar.config` should look like:

``` erlang
{eunit_compile_opts, [
   {src_dirs, ["test"]}
  ]}.
{deps, [
   {lfe, ".*", {git, "git://github.com/rvirding/lfe.git", {tag, "v0.10.0"}}},
   {lutil, ".*", {git, "https://github.com/lfex/lutil.git", {tag, "0.6.7"}}},
   {ltest, ".*", {git, "git://github.com/lfex/ltest.git", {tag, "0.6.4"}}}
  ]}.
{plugins, [
  {'lfe-compile', ".*",
    {git, "https://github.com/lfex/rebar3-lfe-compile.git",
      {branch, "use-new-rebar3-compile"}}}
  ]}.
{provider_hooks, [
  {pre, [{compile, {lfe, compile}}]}
  ]}.
```

We will, of course, eventually be able to use `master` instead of `use-new-rebar3-compile` (but I'll probably encourage folks to use a tagged release instead). I'm going to wait to merge this branch until I figure out how to make LFE compile only once ...

For your dependencies, you'll just need to update the `rebar.config` with the provider hook part:

``` erlang
{provider_hooks, [
  {pre, [{compile, {lfe, compile}}]}
  ]}.
```

I have not tested with any of the following yet:
- Plugins written in LFE
- LFE Plugins with regular LFE dependencies
- LFE Plugins with LFE plugin dependencies
- LFE projects with mixed `.lfe` and `.erl` files
- LFE projects with an `./apps` dir with LFE and/or Erlang projects in them

So, as you can see, still lots to check :-)

**Comment by oubiwann** [comment]
Oh, I found out why LFE is compiling twice -- regular deps and plugins keep their own set of dependencies -- both of which get compiled. I'm not going to try anything clever to get around that. It compiles fast enough so that it's no big deal.

**Comment by timclassic** [comment]
This is great, I hope to do some testing this weekend.

In my environment, I have LFE available system-wide.  Is it required that I include `lfe`, `lutil`, and `ltest` in my dependency list?  Is it currently common for LFE-based dependencies to have their own declared dependency on the `lfe` repo?

Or, perhaps to ask it another way--am I the only person that installs LFE system-wide?

**Comment by oubiwann** [comment]
I've done a little more testing, and the things it was barfing on before is now working: 
- LFE Plugins with regular LFE dependencies
- LFE Plugins with LFE plugin dependencies

This is super-great news and a good sign that things are on the right track ;-)

As for your environment ... you _might_ be the only person I know who is using LFE installed system wide (well, I have a testing setup on my wife's Mac with a Homebrew installation of LFE ...). You should be okay, if you have those libraries installed system-wide, too. It's been a long time since I've tried anything like that, so no guarantees ...

If that's the way you want to run LFE, we should definitely support it so let us know if you run into problems (the LFE mail list is the best place for that).

Let me know how the testing goes for your `./apps` setup with rebar3 ...

**Comment by oubiwann** [comment]
Oh, actually -- let me know in the other ticket --(#2) . I'm going to close this once, since we've got it sorted out.

**Comment by oubiwann** [comment]
I'm going to reopen this, since the core libraries haven't been updated yet. I've done lutil, clj, and kla but haven't gotten to ltest and lcfg yet. Once those two are done and everything's testing well, I'll close this again.

**Comment by oubiwann** [comment]
Hrm, I did end up running across something:
- The `ebin` directory is not getting created during compile (well, not in the top-level dir, that is)
- An `ebin` directory _is_ getting created in the `_build` dir ... and it's an actual dir instead of a symlink, unlike the other dirs.

To be more clear:
- rebar3 compiles the current project as a lib in the build hierarchy, for instance, in the ltest project, there is a `_build/default/lib/ltest` directory.
- the contents of that directory are symlinks back to the current directory
- the exception being the `ebin` directory there

Here's a directory view that illustrates this:

``` bash
 $ ls -al _build/default/lib/clj/
total 12
drwxrwxr-x 3 oubiwann oubiwann 4096 Oct 24 00:46 .
drwxrwxr-x 5 oubiwann oubiwann 4096 Oct 24 00:46 ..
drwxrwxr-x 2 oubiwann oubiwann 4096 Oct 24 00:46 ebin
lrwxrwxrwx 1 oubiwann oubiwann   19 Oct 24 00:46 include -> ../../../../include
lrwxrwxrwx 1 oubiwann oubiwann   16 Oct 24 00:46 priv -> ../../../../priv
lrwxrwxrwx 1 oubiwann oubiwann   16 Oct 24 00:46 test -> ../../../../test
```

For some reason the `ebin` directory is not behaving the same. Need to double-check the plugin code ...

**Comment by oubiwann** [comment]
Okay, got it all sorted out now. A bunch of my LFE libs are running successully on rebar3 now -- any future issues that I discover I'll open a new ticket for. Feel free to do the same!
